Add outline for app-specific commands

sandydays opened the pull request:
Part of #65 

- Includes rough outline of parsing of add, delete, edit, find, and list for patients and appointments.
- Added view command
- Now all these commands operate on the list of persons in the address book. Once appointment and patient specific classes added to model, these commands will only modify those lists. 
- Yet to change tests to fit the changed command parsing methods.
- This skeleton can be adopted for other similar commands like `view-visit`, etc. and modified later.
- Would like feedback on the OOP design too - considering packages for all add commands, all delete commands, and so on...or should it be packages for different models, like appointments, visits, patients, etc.?

crazoter commented on src/main/java/seedu/address/logic/commands/DeleteAppointment.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static final String COMMAND_OBJECT = "appt";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "-" + COMMAND_OBJECT
```
Object then Command

crazoter commented on src/main/java/seedu/address/logic/commands/DeleteAppointment.java:
```diff
@@ -0,0 +1,51 @@
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+public class DeleteAppointment extends DeleteCommand {
```
DeleteAppointmentCommand

crazoter replied:
Note that in the application, the "DeleteCommand" is effectively an "DeletePersonCommand". Please do not be confused and think it is a generic class for deleting items. It is not and thus it should not extend the DeleteCommand (unless it has been explicitly modified to be a generic class, which I doubt so in this case).

crazoter commented on src/main/java/seedu/address/logic/commands/AddPatient.java:
```diff
@@ -0,0 +1,66 @@
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+public class AddPatient extends AddCommand {
```
AddPatientCommand.

crazoter commented on src/main/java/seedu/address/logic/commands/AddCommand.java:
```diff
@@ -19,7 +19,9 @@
     public static final String COMMAND_WORD = "add";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a person to the address book. "
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "-"
```
Note that in the application, the "AddCommand" is effectively an "AddPersonCommand". Please do not be confused and think it is a generic class for adding items. It is not.

crazoter commented on src/main/java/seedu/address/logic/commands/AddAppointment.java:
```diff
@@ -0,0 +1,66 @@
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+public class AddAppointment extends AddCommand {
```
Also, why is AddAppointment extending AddCommand? Note that in the application, the "AddCommand" is effectively an "AddPersonCommand". Please do not be confused and think it is a generic class for adding items. It is not.

crazoter commented on src/main/java/seedu/address/logic/commands/DeleteCommand.java:
```diff
@@ -17,10 +17,11 @@
     public static final String COMMAND_WORD = "delete";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "-"
```
Please remember Object dash Command. If in doubt ask the group before making such changes to avoid unnecessary work like here.

Also note that in the application, the "DeleteCommand" is effectively an "DeletePersonCommand". Please do not be confused and think it is a generic class for adding items. It is not; these commands should be separate.

The message usage should be separate as well. "delete-pat|appt" is messy and hard to read as a user; each command class should only contain message usage for their own command.

crazoter commented on src/main/java/seedu/address/logic/commands/DeletePatient.java:
```diff
@@ -0,0 +1,51 @@
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+public class DeletePatient extends DeleteCommand {
```
Again, Add "Command" behind the name of your class, and don't extend the DeleteCommand (it's a pseudonym for DeletePersonCommand if you have not read the body of the code; it's not a generic delete command class)

crazoter commented on src/main/java/seedu/address/logic/commands/DeletePatient.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static final String COMMAND_OBJECT = "pat";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "-" + COMMAND_OBJECT
```
object - word

crazoter commented on src/main/java/seedu/address/logic/commands/EditAppointment.java:
```diff
@@ -0,0 +1,121 @@
+import seedu.address.model.person.Phone;
+import seedu.address.model.tag.Tag;
+
+public class EditAppointment extends EditCommand {
```
EditAppointmentCommand, and don't extend the EditCommand.

crazoter commented on src/main/java/seedu/address/logic/commands/EditAppointment.java:
```diff
@@ -0,0 +1,121 @@
+
+    public static final String COMMAND_OBJECT = "appt";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "-" + COMMAND_OBJECT
```
object - word

crazoter commented on src/main/java/seedu/address/logic/commands/EditAppointment.java:
```diff
@@ -0,0 +1,121 @@
+            + "by the index number used in the displayed person list. "
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_NAME + "NAME] "
```
Don't forget to update the body and the javadocs

crazoter commented on src/main/java/seedu/address/logic/commands/AddPatient.java:
```diff
@@ -0,0 +1,66 @@
+            + PREFIX_PHONE + "PHONE "
+            + PREFIX_EMAIL + "EMAIL "
+            + PREFIX_ADDRESS + "ADDRESS "
+            + "[" + PREFIX_TAG + "TAG]...\n"
```
I believe you can just remove this class and refactor the current AddCommand class.

sandydays commented:
Hi Matt, thanks for the review and I am sorry if I had to make you read through many lines of code you found quite redundant and messy. I think in some way I was trying to just create some dummy classes that could be modified later and fleshed out properly later, but as I see that perhaps is not a very useful way to go about this, so next time I do put a PR, I will ensure as much as I can from my side that the code is mostly in shape. This was a WIP that was not ready at all to be merged, but to just get some feedback on. Clearly, there were a lot of misunderstandings on my side that surfaced through this PR, so I guess that perhaps was the only good thing out of this.

I would just like to clarify a few things as well based on your comments, if I am wrong on any of them please let me know:
- So it is object-word, like `pat-add` or `appt-add`
- When adding command classes, will ensure to add `Command` to the back of the class names. So I will name the classes as `PatientAddCommand`, `AppointmentDeleteCommand`, etc. 
- You suggested packaging the different commands further. So now, since we are doing object-word, should they be packaged by object (i.e. appointment, patient, etc.) or by the action (add, edit, etc.)? 
- Also, from the AddressBookParser class, do I split different kinds of commands by the action (add, edit, delete) or by object (patient, appointment, etc.)?
- I was in fact planning to use `AddCommand` class as a generic class to be extended by `AddPatient` and `AddAppointment` etc. just that I had not removed all the code within it and cleaned it up. But what is your suggestion on generic classes? Should we have a generic class like `PatientCommand` which all patient related commands extend from? Similarly for appointments etc.?
- Consequently, the changes with MESSAGE_USAGE was just a temporary thing I had done since I hadn't yet cleaned up `AddCommand`, `DeleteCommand`, etc. code yet. 
- Will modify javadocs for any code changes made next time

Again, I am sorry. I believe this could have been quite frustrating and time consuming for you. This was more of an experiment on my side just to get some initial feedback. I will try to give better quality and cleaner code next time. And will try to push in small increments to avoid this. Will start fresh on a new branch as suggested and only focus on appointment commands first perhaps.

crazoter commented:
Hi Sandy, don't worry. Regarding your questions:
* So it is object-word, like pat-add or appt-add
> Yes, `pat-add` and `appt-add`
* When adding command classes, will ensure to add Command to the back of the class names. So I will name the classes as PatientAddCommand, AppointmentDeleteCommand, etc.
> Yes thank you
* You suggested packaging the different commands further. So now, since we are doing object-word, should they be packaged by object (i.e. appointment, patient, etc.) or by the action (add, edit, etc.)?
> When the I use the term "Package", I literally mean Java Packages (see https://www.w3schools.com/java/java_packages.asp) e.g.`package seedu.address.logic.commands`. 
The term "subpackage" is a misuse on my part because each package are just packages with different names (see https://stackoverflow.com/questions/13809713/java-package-in-package), but I was referring to the notion that you have something like `package seedu.address.logic.commands.appointment` like what i've done with visits:
![image](https://user-images.githubusercontent.com/9276766/66826293-bdc17200-ef7e-11e9-9984-2e3d50327711.png)
* Also, from the AddressBookParser class, do I split different kinds of commands by the action (add, edit, delete) or by object (patient, appointment, etc.)?
> Neither, the implementation concept should be kept as is. Treat the command as a whole, because this extraction of entity and action is an additional feature that provides neither the user nor the developer any value at this point. `appt-add` should be recognized as `appt-add`, `appt-delete` should be recognized as `appt-delete` etc, not `appt` and then you search for `add`. What's the point of adding a dash if you're going to implement segregation of commands?
* I was in fact planning to use AddCommand class as a generic class to be extended by AddPatient and AddAppointment etc. just that I had not removed all the code within it and cleaned it up. But what is your suggestion on generic classes? Should we have a generic class like PatientCommand which all patient related commands extend from? Similarly for appointments etc.?
> I believe it's not worth making a generic command for add etc in the first place. What will you extract out? Each entity has different variables, different descriptions, different constraints etc; if you look deeper it's really hard to see similarities on what you can "add". Maybe in v1.4 this can be considered, but at this point I recommend against generic classes.
* Consequently, the changes with MESSAGE_USAGE was just a temporary thing I had done since I hadn't yet cleaned up AddCommand, DeleteCommand, etc. code yet.
* Will modify javadocs for any code changes made next time
> No hurry just something to take note of because it's easy to miss and forget

crazoter commented:
As discussed during the afternoon meeting today, Sandy intends to make her changes on another branch. Closing this PR unless new info crops up
